File: docs/algorithms/code/algorithms/lowestCommonAncestor.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function lowestCommonAncestor(
  root: TreeNode | null,
  p: TreeNode | null,
  q: TreeNode | null
): TreeNode | null {
  if (!root) {
    return null;
  }

  if (root === p || root === q) {
    return root;
  }

  const left = lowestCommonAncestor(root.left, p, q);

  const right = lowestCommonAncestor(root.right, p, q);

  if (left && right) {
    return root;
  }

  return left ? left : right;
}

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function lowestCommonAncestor(
  root: TreeNode | null,
  p: TreeNode | null,
  q: TreeNode | null
): TreeNode | null {
  if (!root) {
    return root;
  }

  if (p.val > root.val && q.val > root.val) {
    return lowestCommonAncestor(root.right, p, q);
  }

  if (p.val < root.val && q.val < root.val) {
    return lowestCommonAncestor(root.left, p, q);
  }

  return root;
}

File: docs/algorithms/code/algorithms/lowestCommonAncestor.go
// 思路：二叉树的公共祖先，思路是递归，先判断root是否为空，如果为空，返回nil，如果root等于p或者q，返回root，然后递归左右子树，如果左右子树都不为空，返回root，如果左子树不为空，返回左子树，如果右子树不为空，返回右子树，如果都为空，返回nil
// 复杂度分析：时间复杂度O(n)，空间复杂度O(h)
// 
// 
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
  if root == nil {
    return nil;
  }

  if root == p || root == q {
    return root;
  }

  leftNode := lowestCommonAncestor(root.Left, p, q);
  rightNode := lowestCommonAncestor(root.Right, p, q);

  if leftNode != nil && rightNode != nil {
    return root;
  }

  if leftNode != nil {
    return leftNode;
  }

  return rightNode;
}